Sources/HTTP/SimpleSocket/SimpleServerSocket.swift

```diff
@@ -0,0 +1,152 @@
+// This source file is part of the Swift.org Server APIs open source project
+//
+// Copyright (c) 2017 Swift Server API project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+//
+
+import Foundation
+import Dispatch
+
+internal class SimpleServerSocket {
+    internal var socketfd: Int32 = -1
+    
+    internal var listeningPort: Int32 = -1
+    
+    internal private(set) var isListening = false
+    
+    internal private(set) var isConnected = false
+
+    internal func socketRead(into readBuffer: inout UnsafeMutablePointer<Int8>, maxLength:Int) throws -> Int {
+        //let readBuffer: UnsafeMutablePointer<CChar> = UnsafeMutablePointer<CChar>.allocate(capacity: maxLength)
+        readBuffer.initialize(to: 0x0)
+        let read = recv(self.socketfd, readBuffer, maxLength, Int32(0))
+//        if read > 0 {
+//            data.append(readBuffer, length: read)
+//        }
+        return read
+    }
+    
+    @discardableResult internal func socketWrite(from buffer: UnsafeRawPointer, bufSize: Int) throws -> Int {
+        return send(self.socketfd, buffer, Int(bufSize), Int32(0))
+    }
+    
+    internal func shutdownAndClose() {
+        //print("Shutting down socket \(self.socketfd)")
+        if self.isListening || self.isConnected {
+            //print("Shutting down socket")
+            _ = shutdown(self.socketfd, Int32(SHUT_RDWR))
+            self.isListening = false
+        }
+        self.isConnected = false
+        close(self.socketfd)
+    }
+        
+    internal func acceptClientConnection() throws -> SimpleServerSocket {
+        let retVal = SimpleServerSocket()
```

Comment:
Call this `socket`?

Reply:
It's not, though.  It only implements a small subset of the `socket(2)` APIs. Which is all we need for this application, but it's not a general purpose `Socket` object.